lib/circbufwriter/writer.go

```diff
@@ -0,0 +1,112 @@
+package circbufwriter
+
+import (
+	"io"
+	"sync"
+	"time"
+
+	"github.com/armon/circbuf"
+)
+
+type circbufWriter struct {
+	// circle buffer for data to write
+	buf *circbuf.Buffer
+
+	// error to return from the writer
+	err error
+
+	// bufLock syncronizes access to err and buf
+	bufLock sync.Mutex
+
+	// wrapped writer
+	wr io.Writer
+
+	// signals to flush the buffer
+	flushCh chan struct{}
+}
+
+// New created a circle buffered writer that wraps the given writer. The
+// bufferSize is the amount of data that will be stored in memory before
+// overwriting.
+func New(w io.Writer, bufferSize int64) io.WriteCloser {
+	buf, _ := circbuf.NewBuffer(bufferSize)
+	c := &circbufWriter{
+		buf:     buf,
+		wr:      w,
+		flushCh: make(chan struct{}, 1),
+	}
+	go c.flushLoop()
+	return c
+}
+
+// Write will write the data to the buffer and attempt to flush the buffer to
+// the wrapped writer. If the wrapped writer blocks on write, subsequent write
+// will be written to the circle buffer.
+func (c *circbufWriter) Write(p []byte) (nn int, err error) {
+	// If the last write returned an error, return it here
+	c.bufLock.Lock()
+	if c.err != nil {
+		return nn, c.err
+	}
+
+	// Write to the buffer
+	nn, err = c.buf.Write(p)
+	c.bufLock.Unlock()
+
+	// Signal to flush the buffer
+	select {
+	case c.flushCh <- struct{}{}:
+	default:
+		// flush is blocked
+	}
+	return nn, err
+}
+
+func (c *circbufWriter) Close() error {
+	// Guard against double closing channel
+	select {
+	case <-c.flushCh:
+	default:
+		close(c.flushCh)
+	}
+
+	// if the last write errored, it will return here
+	c.bufLock.Lock()
+	defer c.bufLock.Unlock()
+	return c.err
+}
+
+func (c *circbufWriter) flushLoop() {
+	// Check buffer every 100ms in case a flush from Write was missed
+	ticker := time.NewTicker(time.Millisecond * 100)
+	for {
+		var err error
+		select {
+		case _, ok := <-c.flushCh:
+			if !ok {
+				// Close called, exiting loop
+				return
+			}
+			err = c.flush()
+		case <-ticker.C:
+			err = c.flush()
+		}
+
+		c.bufLock.Lock()
+		c.err = err
```

Comment:
This is good enough IMO - but still risks overwriting errors with nils; in reality, probably errors will be persistent across calls.  And given that it's a fixed circular buffer, there is no guarantee that everything is written, so dropping some content because the buffer is full, or because the underlying writer transiently failed shouldn't be dramatically different.

I would probably add a comment stating that this is a "best-effort" writer or something and call it a day tbh.

Reply:
Added a comment about the returned error in Write